File: vanilla-mc-leaderboard.py
#vanilla-mc-leaderboard.py
#Copyright (C) 2020 Alexander Theulings, ketchupcomputing.com <alexander@theulings.com>
#
#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
# 
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
# 
#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 
#Documentation for this file can be found at https://ketchupcomputing.com/projects/mc-leaderboard/
#

import datetime
import nbt
import os
import shutil
import json
import time
from ftplib import FTP
import mysql.connector

class itemPair:
    def __init__(self, setMcItemName, setValue):
        self.mcItemName = setMcItemName
        self.value = setValue

class itemSet:
    def __init__(self, setStoreId, setMinimumCount, setCheckInvent, setCheckEnders, setCheckShulkers, setPairs):
        self.storeId = setStoreId
        self.minimumCount = setMinimumCount
        self.checkInvent = setCheckInvent
        self.checkEnders = setCheckEnders
        self.checkShulkers = setCheckShulkers
        self.pairs = setPairs

def scanStorageFor(storage, forItemName,  checkShulkers):
    count = 0
    for slot in storage:
        if str(slot["id"]) == forItemName:
            count += int(str(slot["Count"]))
        if checkShulkers and str(slot["id"]) == "minecraft:shulker_box":
            if "tag" in slot:
                count += scanStorageFor(slot["tag"]["BlockEntityTag"]["Items"], forItemName, checkShulkers)
    return count

with open('vanilla-mc-leaderboard-config.json') as json_file:
    data = json.load(json_file)

    #Minecraft server settings
    ftpAddr = data['ftp']["address"]
    ftpUname = data['ftp']["username"]
    ftpPword = data['ftp']["password"]
    ftpUserDir = data['ftp']["user_directory"]
    #Output to SQL settings
    sqlAddr = data["sql"]["address"]
    sqlUname = data["sql"]["username"]
    sqlPword = data["sql"]["password"]
    sqlDB = data["sql"]["name"]
    sqlMainTableName = data["sql"]["main_table_name"]
    #Directory Settings
    localWorkingDir = data["working_directory"]["path"]
    dirPrefix = data["working_directory"]["prefix"]
    #Score record settings
    keepScoreRecords = data["score_records"]["keep_score_records"]
    scoreRecordId = data["score_records"]["store_id"]
    minimumScoreThreshold = data["score_records"]["minimum_score"]

    itemRecordSqlTable = data["item_record_sql_table_name"]
    itemSets = []
    for record in data["item_record_sets"]:
        itemPairs = []
        for pair in record["item_pairs"]:
            itemPairs.append(itemPair(pair["name"], pair["value"]))
        itemSets.append(itemSet(record["store_id"], record["minimum_count"], record["check_inventory"], record["check_ender_chest"], record["check_shulkers"], itemPairs))

#Create a directory to work in
fullWorkingDir = localWorkingDir + dirPrefix + str(time.time())
os.mkdir(fullWorkingDir)
os.chdir(fullWorkingDir)

#Connect to Minecraft server
ftp = FTP(ftpAddr)
ftp.login(ftpUname, ftpPword)
ftp.cwd(ftpUserDir)
fileList = ftp.nlst()

#Connect to database if enabled
connection = mysql.connector.connect(host=sqlAddr, database=sqlDB, user=sqlUname, password=sqlPword)
cursor = connection.cursor(buffered=True)

today = datetime.datetime.now().strftime("%Y-%m-%d")
fullDate = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")

#Download player files and extract scores
for listing in fileList:
    if listing.endswith(".dat_old"):
        continue
    with open(listing, 'wb') as fp:
        ftp.retrbinary('RETR ' + listing, fp.write)
    nbtfile = nbt.nbt.NBTFile(listing,'rb')
    if keepScoreRecords:
        if int(str(nbtfile["Score"])) >= minimumScoreThreshold:
            cursor.execute("INSERT INTO " + sqlMainTableName + " (script_id, uuid, count, time) VALUES (%s, %s, %s, %s)", [scoreRecordId, listing.replace(".dat", ""), int(str(nbtfile["Score"])), fullDate])

    for itemSet in itemSets:
        itemCount = 0
        for itemPair in itemSet.pairs:
            if itemSet.checkInvent:
                itemCount += scanStorageFor(nbtfile["Inventory"], itemPair.mcItemName, itemSet.checkShulkers) * itemPair.value
            if itemSet.checkEnders:
                itemCount += scanStorageFor(nbtfile["EnderItems"], itemPair.mcItemName, itemSet.checkShulkers) * itemPair.value

        if itemCount >= itemSet.minimumCount:
            cursor.execute("INSERT INTO " + sqlMainTableName + " (script_id, uuid, count, time) VALUES (%s, %s, %s, %s)", [itemSet.storeId, listing.replace(".dat", ""), itemCount, fullDate])

connection.commit()
cursor.close()
connection.close()

#Delete working directory
os.chdir("..")
shutil.rmtree(fullWorkingDir)

File: vanilla-mc-leaderboard.php
<?php
//vanilla-mc-leaderboard.php
//Copyright (C) 2020 Alexander Theulings, ketchupcomputing.com <alexander@theulings.com>
//
//This program is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//(at your option) any later version.
// 
//This program is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
// 
//You should have received a copy of the GNU General Public License
//along with this program.  If not, see <http://www.gnu.org/licenses/>.
// 
//Documentation for this file can be found at https://ketchupcomputing.com/projects/mc-leaderboard/
//
//This script displays scores parsed by vanilla-mc-leaderboard.py

class vanillaMcLeaderboard{
    //Settings
    private static $sqlAddr = "localhost";
    private static $sqlUname = "database.username";
    private static $sqlPword = "database.password";
    private static $sqlDB = "database.name";
    private static $sqlMainTableName = "database.table.name";

    private static $mcUnamesTable = "database.table.name";
    private static $mcUnamesCacheFor = 168;

    private static $htmlPre = "Records as of %t UTC<br><hr><br>";
    private static $htmlRow = "%p: %u - %c<BR>";
    private static $htmlPost = "";

    private static $displayPoweredBy = true;

    public static function display($displayStoreId = 0, $resultCount = 25, $timestamp = 0, $useExternalConfig = null){
        //Load config from file
        if ($useExternalConfig != null){
            $configFile = fopen($useExternalConfig, "r");
            $configJson = fread($configFile, filesize($useExternalConfig));
            fclose($configFile);
            $config = json_decode($configJson, true);

            self::$sqlAddr = $config["sql"]["address"];
            self::$sqlUname = $config["sql"]["username"];
            self::$sqlPword = $config["sql"]["password"];
            self::$sqlDB = $config["sql"]["name"];
            self::$sqlMainTableName = $config["sql"]["main_table_name"];

            self::$mcUnamesTable = $config["username_records"]["sql_table_name"];

            self::$htmlPre = $config["html_output"]["pre"];
            self::$htmlRow = $config["html_output"]["row"];
            self::$htmlPost = $config["html_output"]["post"];
            self::$displayPoweredBy = $config["html_output"]["display_powered_by"];
        }

        //Connect to db
        $dbConn = new mysqli(self::$sqlAddr, self::$sqlUname, self::$sqlPword, self::$sqlDB);
        if ($dbConn->connect_error){
            echo "vanilla-mc-leaderboard encountered an error. See site log for more details.";
            error_log("vanilla-mc-leaderboard error connecting to database: " . $dbConn->connect_error);
            return false;
        }

        //If no timestamp has been requested get the last inserted timestamp from the database
        if($timestamp == 0){
            $stmt = $dbConn->prepare("SELECT `time` FROM " . self::$sqlMainTableName . " WHERE `store_id`=? ORDER BY `id` DESC LIMIT 1");
            $stmt->bind_param("i", $displayStoreId);
            $stmt->execute();
            $timestamp = $stmt->get_result()->fetch_assoc()["time"];
            $stmt->close();
        }
        
        //Get leaderboard
        $stmt = $dbConn->prepare("SELECT `uuid`, count FROM " . self::$sqlMainTableName . " WHERE `time`=? AND `store_id`=? ORDER BY `count` DESC LIMIT ?");
        $stmt->bind_param("sii", $timestamp, $displayStoreId, $resultCount);
        $stmt->execute();
        $results = $stmt->get_result();

        //Display leaderboard
        echo str_replace("%t", $timestamp, self::$htmlPre);
        $pos = 1;
        while($row = $results->fetch_assoc()){
            $rowStr = str_replace("%p", $pos, self::$htmlRow);
            $rowStr = str_replace("%u", self::getUsername($row["uuid"]), $rowStr);
            $rowStr = str_replace("%c", $row["count"], $rowStr);
            echo $rowStr;
            $pos ++;
        }
        $stmt->close();
        $dbConn->close();
        echo self::$htmlPost;

        if (self::$displayPoweredBy){
            echo "<br><p style=\"font-size: small\">Powered by <a href=\"https://ketchupcomputing.com/docs/mc-leaderboard\" target=\"_blank\">vanilla-mc-leaderboard</a>.</p>";
        }
    }

    //Get name from database if cached
    private static function getUsername($uuid){
        //Connect to db
        $dbConn = new mysqli(self::$sqlAddr, self::$sqlUname, self::$sqlPword, self::$sqlDB);
        if ($dbConn->connect_error){
            die("Connection to database failed - " . $dbConn->connect_error);
        }

        $stmt = $dbConn->prepare("SELECT `uname` FROM " . self::$mcUnamesTable . " WHERE `uuid`=? LIMIT 1");
        $stmt->bind_param("s", $uuid);
        $stmt->execute();
        $results = $stmt->get_result();
        if($results->num_rows == 0){
            //Not cached, get it from Mojang
            $stmt->close();
            $dbConn->close();
            return self::retrieveUsername($uuid);
        }else{
            $row = $results->fetch_assoc();
            $uname = $row["uname"];
            $stmt->close();
            $dbConn->close();
            return $uname;
        }
    }

    //Gets username from Mojang and caches it
    private static function retrieveUsername($uuid){
        //Connect to db
        $dbConn = new mysqli(self::$sqlAddr, self::$sqlUname, self::$sqlPword, self::$sqlDB);
        if ($dbConn->connect_error){
            die("Connection to database failed - " . $dbConn->connect_error);
        }
       
        //Remove old records of uuid
        $stmt = $dbConn->prepare("DELETE FROM " . self::$mcUnamesTable . " WHERE `uuid`=?");
        $stmt->bind_param("s", $uuid);
        $stmt->execute();
        $stmt->close();

        //Get username from Mojang
        $cutUuid = str_replace ("-", "", $uuid);
        $unameData = file_get_contents("https://api.mojang.com/user/profiles/" . $cutUuid . "/names");
        $unameJson = json_decode($unameData);
        $uname = $unameJson[count($unameJson) - 1]->{"name"};
        
        //Store username in database
        $stmt = $dbConn->prepare("INSERT INTO " . self::$mcUnamesTable . " (uuid, uname) VALUES (?, ?)");
        $stmt->bind_param("ss", $uuid, $uname);
        $stmt->execute();
        $stmt->close();
        $dbConn->close();
        return $uname;
    }
}
